components/common-go/grpc/ratelimit.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) 2020 TypeFox GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package grpc
+
+import (
+	"context"
+	"time"
+
+	"github.com/gitpod-io/gitpod/common-go/util"
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// RateLimit configures the reate limit for a function
+type RateLimit struct {
+	Block          bool          `json:"block"`
+	BucketSize     uint          `json:"bucketSize"`
+	RefillInterval util.Duration `json:"refillInterval"`
+}
+
+// NewRatelimitingInterceptor creates a new rate limiting interceptor
+func NewRatelimitingInterceptor(f map[string]RateLimit) RatelimitingInterceptor {
+	funcs := make(map[string]ratelimitedFunction, len(f))
+	for name, fnc := range f {
+		funcs[name] = ratelimitedFunction{
+			Block: fnc.Block,
+			L:     rate.NewLimiter(rate.Every(time.Duration(fnc.RefillInterval)), int(fnc.BucketSize)),
+		}
+	}
+	return funcs
+}
+
+// RatelimitingInterceptor limits how often a gRPC function may be called. If the limit has been
+// exceeded, we'll return resource exahusted.
+type RatelimitingInterceptor map[string]ratelimitedFunction
+
+type ratelimitedFunction struct {
+	Block bool
+	L     *rate.Limiter
+}
+
+// UnaryInterceptor creates a unary interceptor that implements the rate limiting
+func (r RatelimitingInterceptor) UnaryInterceptor() grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
+		f, ok := r[info.FullMethod]
+		if ok {
+			if f.Block {
+				err := f.L.Wait(ctx)
+				if err == context.Canceled || err == context.DeadlineExceeded {
+					return nil, err
+				}
+				if err != nil {
+					return nil, status.Error(codes.ResourceExhausted, err.Error())
+				}
+			} else if !f.L.Allow() {
+				return nil, status.Error(codes.ResourceExhausted, "too many requests")
+			}
+		}
+
+		return handler(ctx, req)
```

Comment:
Just a suggestion to improve code readability and maybe prevent some unnecessary computation 

I think it's always good to remove an indentation level when it's possible :slightly_smiling_face: 

```suggestion
		f, ok := r[info.FullMethod];
		if !ok {
			return handler(ctx, req)
		}

		if f.Block {
			err := f.L.Wait(ctx); 
			if err == context.Canceled || err == context.DeadlineExceeded {
				return nil, err
			}
		  	if err != nil {
		  		return nil, status.Error(codes.ResourceExhausted, err.Error())
		  	}
		} else if !f.L.Allow() {
			return nil, status.Error(codes.ResourceExhausted, "too many requests")
		}
		return handler(ctx, req)
```

Reply:
[Aligning the happy path on the left](https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88) is good style indeed - and the vast majority of our Go code follows this advice. As always, there are exceptions though, much like in this case.

Calling `handler` is something that we must do once, and once only. Also, the nature of this function is something that "stands between the request and the actual call". With a single **short** `if ok { ... }` block it's easier see this behaviour straight away. Put differently, I found the version that called `handler` twice, harder to follow.

Note: it's absolutely key that the `ok` block is short and itself doesn't indent much. If that were any different the above argument wouldn't hold anymore.